### ISSUE TITLE
Interoperability with PyVista

### ISSUE THREAD

**Issue by banesullivan** [gimli-org/gimli#178]
Let's bring [PyVista](https://docs.pyvista.org) to PyGIMLi! Inspired by @florian-wagner's comment: https://github.com/pyvista/pyvista/issues/208#issuecomment-490545903

> Next pyGIMLi release will ship with a vista-based 3D viewer.

Somewhat related to #15 

I'd like to use this issue as a place to outline how bringing PyVista into will play out.

## The Plan

I've been poking around the source code lately trying to get a feel for how this will play out

### Proof of Concept

I'm thinking we could start by implementing a routine just like what's currently in [mayaview.py](https://github.com/gimli-org/gimli/blob/master/python/pygimli/viewer/mayaview.py) where we could export the mesh to a VTK file then load it using PyVista and render it, but I don't think that's a very elegant solution.

### Moving Towards an Interface between VTK and GIMLi

To move towards a more elegant solution, I'm thinking we should create an interface between GIMLi mesh types and VTK mesh types on the Python side via PyVista - PyVista provides an interface that directly inherits from VTK's types and provides a lot of convenience methods for generating VTK mesh types via NumPy arrays (it's far better streamlined than what Mayavi offers in my not-so-humble opinion). This interface would allow users to go back and forth between PyVista/VTK and GIMLi mesh types - thus expanding the tools available for creating meshes and performing mesh analysis/filtering.

#### How would this work?

I'm not really sure... hoping you all can help here! 

I'm mainly unsure because PyGIMLi interfaces back to C++ code... do you all have direct bindings for the GIMLi mesh types?

#### What might this look like?

I'm thinking there might be some overlap with what I implemented for SimPEG's `discretize` with a VTK mixin

**Comment by florian-wagner** [comment]
Hey Bane,

thanks for triggering, but we are already on it: http://www2.geo.uni-bonn.de/~wagner/pg/_examples_auto/index.html

We are developing pygimli 1.1 in the fw_cleaning Branch. In interactive Mode, @frodo4fingers made a qt window around the pg.show command. Right now it's more a proof of concept. We are exporting a VTK file to a tempfile and reload it. That should Work without file Export in the future.

Great Software!

**Comment by florian-wagner** [comment]
https://github.com/gimli-org/gimli/blob/fw_cleaning/python/pygimli/viewer/vtkiview.py

Thats what we have so far. The ExportVTK is in C++ and writing to a file, which seems to perform better as iterating in python over mesh entities. An elegant and performant interface would definitely be interesting and we appreciate your help.

Best wishes
Florian

**Comment by banesullivan** [comment]
This is awesome! I didn't realize you all were already on top of this.

You all are really taking advantage of PyVista - super cool examples! :hey this:

**Comment by banesullivan** [comment]
PyGIMLi fully supports tetrahedral meshes, correct?

Perhaps [`tetgen`](http://tetgen.pyvista.org) might also be useful? It's 3D tetrahedral mesh generator built on top of PyVista and the C++ TetGen library to make meshes out of any closed surface:

![cow](https://user-images.githubusercontent.com/22067021/58358623-bb367280-7e4d-11e9-8228-6c0631b2f2b9.gif)

I don't imagine a cow is a great example here, but perhaps there are other closed surfaces of objects that would make more sense to run a simulation on?

**Comment by carsten-forty2** [comment]
Indeed .. tetgen is a nice piece of software .. we apply it as 3d mesh generator since 2006. 

Recent example from Rochlitz et.al 2018:

![rochlitz2018](https://user-images.githubusercontent.com/6043771/58360968-e770e400-7e8b-11e9-8ea3-15676815114c.png)

We made a basic toolset to create/operate with the input PLCs. 
Currently, these tools are only accessible through system calls to a prior version of gimli but they will get a wrapper to the pg.meshtools or will be migrated to native python code in the upcoming 1.1. version. We will definitely take a look at the pyvista tetgen wrapper if we can share synergies.

**Comment by banesullivan** [comment]
Woah, that's really awesome!

If you all think PyVista's `tetgen` could meet GIMLi's needs then I'd be happy let `tetgen` be a point of synergy between @pyvista and @gimli-org! 

As TetGen is incredibly general-use, it might make sense to maintain it under @pyvista and have GIMLi call back to PyVista's `tetgen` so we can reach a wider audience with regards to maintenance

**Comment by florian-wagner** [comment]
Hi @banesullivan,

that sounds generally interesting. Can you give us a quick overview on what the pyvista tetgen wrapper is capable of right now and what is on the roadmap? 

The tetgen call for a given piecewise linear complex (PLC) is straightforward, the main challenge is in creating the PLC with user-friendly high-level functions. Are you planning advanced functionalites in this direction? e.g. boolean operations on geometrical entities, including topography, etc. 

This example uses `pg.meshtools.createCube` and a subsequent tetgen call http://www2.geo.uni-bonn.de/~wagner/pg/_examples_auto/dc_and_ip/plot_modTank3d.html#sphx-glr-examples-auto-dc-and-ip-plot-modtank3d-py

Best wishes
Florian

**Comment by banesullivan** [comment]
I totally, dropped the ball on this one - thanks for reminding me, @florian-wagner!

> Can you give us a quick overview on what the pyvista tetgen wrapper is capable of right now and what is on the roadmap?

At first, I waited to follow up because there are some serious licensing concerns, and I wanted to consult the makers of TetGen. Unfortunately, TetGen is under a GPLv3 license which means PyVista cannot have *any* interaction with TetGen. However, our TetGen wrapper can interact with PyVista all day because the wrapper is under GPLv3 and PyVista is under MIT - we can go one way but not the other. Unfortunately, if even a single line of code in PyGIMLi interacts with TetGen or our TetGen wrapper, you'd have to relicense this entire project under GPLv3 (to my knowledge), which is something I would approach with the utmost caution.

The roadmap for PyVista and our TetGen wrapper: The wrapper will grow as we have time/see a need for it and we will make sure it is fully compatible with PyVista. With that said, we will not add any code in PyVista to send meshes to the TetGen wrapper or leverage the wrapper in any way - we'll leave this to the user (which only takes an extra line or two of code).

> The tetgen call for a given piecewise linear complex (PLC) is straightforward, the main challenge is in creating the PLC with user-friendly high-level functions. Are you planning advanced functionalites in this direction? e.g. boolean operations on geometrical entities, including topography, etc.

I think so... the TetGen wrapper's purpose is to make those kinds of calls high-level, user-friendly, and support PyVista mesh types. I imagine all of this is possible if we collaborate, though @akaszynski is the actual author of TetGen so maybe he could chime in? As for other mesh tools like boolean operations, etc. - PyVista is already built for that!

**Comment by carsten-forty2** [comment]
Thank you for your thought-provoking answer, indeed we are aware of TetGen's licence (AGPLv3) and had a couple internal discussions in the past. 

With my interpretation of the GPL, we are on the safe side but I can be wrong since I'm no lawyer. We do not ship a modified copy of TetGen, we don't link TetGen in any form and we don't import any TetGen related wrapper. The only thing we do is creating a proper PLC and a command line to make a system call to the TetGen binary. Its in the users responsibility to provide the TetGen binary somewhere in the system path.  It would be much nicer to communicate directly with the TetGen library,  but like you say .. this would give us GPL problems. 

With no linking, no usage of shared memory and no extensive data exchange, we interpret 
[#GPLPlugins](http://www.gnu.org/licenses/gpl-faq.html#GPLPlugins) in a way that pygimli and TetGen stays separate programs. But maybe its more easy since we do not distribute or convey TetGen. 
[#NoDistributionRequirements](http://www.gnu.org/licenses/gpl-faq.html#NoDistributionRequirements)

I didn't recognise the authorship of TetGen has been changed: http://wias-berlin.de/software/index.jsp?id=TetGen&lang=1#Introduction. He seems quite busy .. I tried a bug report some months ago but got no answer. Did you got any answer from the author?
